### DISCUSSION TITLE
Unsure how Keyclock redirect URI changes to port 3000 of Outline

### DISCUSSION THREAD

**Discussion by TheDanHealy** [outline/outline#6002] [Self Hosting] [answered]
Outline: Latest
Keycloak: 22.0.4

I have both Outline and Keycloak (previously deployed and in-use in my environment) behind an Nginx SSL terminating reverse proxy. They're available at outline.example.com and keycloak.example.com both inside and outside the network. 

When I configure Keycloak's redirect URI to `https://outline.example.com:443/*`, the resulting URL in the browser (which fails) is `https://outline.example.com:3000/?notice=state-mismatch`
 
Here's a snippet from my compose environment variables:

```
      OIDC_CLIENT_ID: outline
      OIDC_CLIENT_SECRET: my_random_key
      OIDC_AUTH_URI: https://keycloak.example.com/realms/myrealm/protocol/openid-connect/auth
      OIDC_TOKEN_URI: https://keycloak.example.com/realms/myrealm/protocol/openid-connect/token
      OIDC_USERINFO_URI: https://keycloak.example.com/realms/myrealm/protocol/openid-connect/userinfo

      # Specify which claims to derive user information from
      # Supports any valid JSON path with the JWT payload
      OIDC_USERNAME_CLAIM: preferred_username

      # Display name for OIDC authentication
      OIDC_DISPLAY_NAME: Keycloak

      # Space separated auth scopes.
      OIDC_SCOPES: openid profile email

      REDIS_URL: redis://redis:6379
      URL: https://outline.example.com
      PORT: 3000
```
![image](https://github.com/outline/outline/assets/43013393/80e9e98c-cf3b-4e52-8cdb-7e0741ee9d7e)

**Comment by TheDanHealy** [comment]
Also, the outline app logs produced this, but unsure if this affects the weird port redirect.

```
{"error":"State not return in OAuth flow","level":"error","message":"Error during authentication","stack":"BadRequestError: State not return in OAuth flow\n    at OAuthStateMismatchError (/opt/outline/build/server/errors.js:122:34)\n    at StateStore.verify (/opt/outline/build/server/utils/passport.js:46:61)\n    at OAuth2Strategy.authenticate (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:222:26)\n    at _passportOauth.Strategy.authenticate (/opt/outline/build/plugins/oidc/server/auth/oidc.js:39:26)\n    at attempt (/opt/outline/node_modules/passport/lib/middleware/authenticate.js:369:16)\n    at authenticate (/opt/outline/node_modules/passport/lib/middleware/authenticate.js:370:7)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:194:7\n    at new Promise (<anonymous>)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:193:12\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:143:7\n    at new Promise (<anonymous>)\n    at passportAuthenticate (/opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:107:15)\n    at passportMiddleware (/opt/outline/build/server/middlewares/passport.js:75:7)\n    at /opt/outline/node_modules/dd-trace/packages/datadog-instrumentations/src/koa.js:88:57\n    at passportMiddleware (/opt/outline/node_modules/dd-trace/packages/datadog-shimmer/src/shimmer.js:26:21)\n    at dispatch (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)\n    at /opt/outline/node_modules/koa-router/lib/router.js:346:16\n    at dispatch (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:60:14\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
```

**Comment by TheDanHealy** [comment]
I think this is related: https://github.com/outline/outline/discussions/5388

**Comment by TheDanHealy** [comment]
Any hints or pointers? I'm unsure where to go from here except to dig into the Outline code. I was really hoping this would be an easy deployment into an existing Docker Swarm cluster.

**Reply by tommoor** [reply]
Not sure, but why are you explicitly adding `:443` everywhere, it's a given with `https` so you should remove that from everywhere it's configured.

**Reply by TheDanHealy** [reply]
Yeah, you're right. Those port numbers don't exist in my config anymore. I suppose my screenshots above are a little old. I'll update. But I'm still experiencing the same issue sadly.

**Reply by tommoor** [reply]
I'd suggest going through the config line by line, I'm confident this is a configuration issue. Never had anything like this reported before, and Outline doesn't really have any control over where Keycloak redirects to post-login.

**Comment by TheDanHealy** [comment, marked as answer]
I've found the fix. There were a few `proxy_set_header` lines missing from my nginx config for this site that made the difference. I compared my config with https://docs.getoutline.com/s/hosting/doc/nginx-6htaRboR57 and merged the changes. After that, it worked.
